src/Core/Hooks/Model/InstanceManager.php

```diff
@@ -0,0 +1,104 @@
+<?php
+/**
+ * @copyright Copyright (C) 2010-2023, the Friendica project
+ *
+ * @license GNU AGPL version 3 or any later version
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ */
+
+namespace Friendica\Core\Hooks\Model;
+
+use Dice\Dice;
+use Friendica\Core\Hooks\Capabilities\IAmAStrategy;
+use Friendica\Core\Hooks\Capabilities\ICanManageInstances;
+use Friendica\Core\Hooks\Exceptions\HookInstanceException;
+use Friendica\Core\Hooks\Exceptions\HookRegisterArgumentException;
+
+/** {@inheritDoc} */
+class InstanceManager implements ICanManageInstances
+{
+	protected $instance          = [];
+	protected $instanceArguments = [];
+	protected $decorator         = [];
+
+	/** @var Dice */
+	protected $dice;
+
+	public function __construct(Dice $dice)
+	{
+		$this->dice = $dice;
+	}
+
+	/** {@inheritDoc} */
+	public function registerStrategy(string $interface, string $name, string $class, array $arguments = null): ICanManageInstances
+	{
+		if (!is_a($class, $interface, true)) {
+			throw new HookRegisterArgumentException(sprintf('%s is not a valid class for the interface %s', $class, $interface));
+		}
+
+		if (!is_a($class, IAmAStrategy::class, true)) {
+			throw new HookRegisterArgumentException(sprintf('%s does not inherit from the marker interface %s', $class, IAmAStrategy::class));
+		}
+
+		if (!empty($this->instance[$interface][$name])) {
+			throw new HookRegisterArgumentException(sprintf('A class with the name %s is already set for the interface %s', $name, $interface));
+		}
```

Comment:
Doesn't this mean addons can't override the default logger instantiation strategy?

Reply:
An addon or any other strategies must not override the possible strategies, it adds another strategy --> so it must not override existing unique keys for it.

The concrete addon instance is chosen with `getInstance($class, 'addon_name')`.

Reply:
Ok. I'd like to see a full example of this to wrap my head around it.

Reply:
Here's how we register new strategies or decorators durin the load:
https://github.com/friendica/friendica/pull/12680/files#diff-015c493c07a82470af3995c977c6fee05db967b28d956dedaf3722685f3f05cbR54-R60

And here's the way we use it 
https://github.com/friendica/friendica/pull/12680/files#diff-015c493c07a82470af3995c977c6fee05db967b28d956dedaf3722685f3f05cbR81

If the `$name`, loaded from the config, is `'monolog'`, it would search for a registered strategy with the name `monolog`. So the addon has to register it's strategy first at the addon directory with `$this->instanceManager		->registerStrategy(LoggerInterface::class, 'monolog', MonologLogger::class)

Reply:
Oh, and the `system.logger_config` controls the name, I get it now!

And we can list the strategies in the admin config the same way we do for FileSystem drivers.